openSearchLane.go

```diff
@@ -0,0 +1,425 @@
+	}
+	if config.BackOffLimit == 0 {
+		osl.openSearchConnection.config.BackOffLimit = DefaultBackOffLimit
+	}
+
+	go osl.openSearchConnection.processConnection(osl)
+
+	return
+
+}
+
+func (osl *openSearchLane) Derive() Lane {
+	ll := deriveLogLane(&osl.logLane, context.WithValue(osl.Context, parent_lane_id, osl.LaneId()), osl.tees, osl.cr)
+
+	osl2 := openSearchLane{
+		openSearchConnection: osl.openSearchConnection,
+	}
+	osl.openSearchConnection.attach()
+
+	ll.clone(&osl2.logLane)
+	osl2.logLane.writer = log.New(&osl2, "", 0)
+
+	return &osl2
+}
+
+func (osl *openSearchLane) DeriveWithCancel() (Lane, context.CancelFunc) {
+	childCtx, cancelFn := context.WithCancel(context.WithValue(osl.logLane.Context, parent_lane_id, osl.logLane.LaneId()))
+	ll := deriveLogLane(&osl.logLane, childCtx, osl.tees, osl.cr)
+
+	osl2 := openSearchLane{
+		openSearchConnection: osl.openSearchConnection,
+	}
+	osl.openSearchConnection.attach()
+
+	ll.clone(&osl2.logLane)
+	osl2.logLane.writer = log.New(&osl2, "", 0)
+
+	return &osl2, cancelFn
+}
+
+func (osl *openSearchLane) DeriveReplaceContext(ctx context.Context) Lane {
+	ll := deriveLogLane(&osl.logLane, context.WithValue(ctx, parent_lane_id, osl.LaneId()), osl.tees, osl.cr)
+
+	osl2 := openSearchLane{
+		openSearchConnection: osl.openSearchConnection,
+	}
+	osl.openSearchConnection.attach()
+
+	ll.clone(&osl2.logLane)
+	osl2.logLane.writer = log.New(&osl2, "", 0)
+
+	return &osl2
+}
+
+func (osl *openSearchLane) SetMetadata(key string, val string) {
+	osl.openSearchConnection.mu.Lock()
+	defer osl.openSearchConnection.mu.Unlock()
+
+	if osl.metadata == nil {
+		osl.metadata = make(map[string]string)
+		osl.metadata["timestamp"] = time.Now().UTC().Format(time.RFC3339)
+	}
+
+	osl.metadata[key] = val
+}
+
+func (osl *openSearchLane) Close() {
+	osl.openSearchConnection.detach()
+}
+
+func (osl *openSearchLane) Write(p []byte) (n int, err error) {
+	osl.openSearchConnection.mu.Lock()
+	defer osl.openSearchConnection.mu.Unlock()
+
+	logEntry := string(p)
+
+	if strings.Contains(logEntry, "\n") {
+		logEntry = strings.ReplaceAll(logEntry, "\n", "")
+	}
+
+	// TODO remove this line
+	fmt.Println(logEntry)
+
+	parentLaneId, ok := osl.Context.Value(parent_lane_id).(string)
+	if !ok {
+		parentLaneId = ""
+	}
+
+	logData := &OpenSearchLogMessage{
+		AppName:      osl.openSearchConnection.config.OpenSearchAppName,
+		ParentLaneId: parentLaneId,
+		JourneyID:    osl.journeyId,
+		LaneID:       osl.LaneId(),
+		LogMessage:   logEntry,
+		Metadata:     osl.metadata,
+	}
+
+	osl.openSearchConnection.logBuffer = append(osl.openSearchConnection.logBuffer, logData)
+
+	osl.openSearchConnection.messagesQueued++
+
+	if len(osl.openSearchConnection.logBuffer)%osl.openSearchConnection.config.LogThreshold == 0 {
+		osl.openSearchConnection.wakeCh <- struct{}{}
+	}
+
+	return len(p), nil
+}
+
+func (osl *openSearchLane) Stats() {
+	osl.openSearchConnection.mu.Lock()
```

Comment:
It's strange that responsibilities belonging to openSearchConnection are done in the implementation of openSearchLane.

Reply:
They share the same Mutex and some of the values of openSearchConnection, that's why I'm using the same mutex

Reply:
Usually it would look like this:

```
func (osl *openSearchLane) Stats() OslStats {
    return osl.openSearchConnection.Stats()
}

func (osc *openSearchConnection) Stats() OslStats {
    osc.mu.Lock()
    defer osc.mu.Unlock()

    return OslStats{
       x: osc.x,
       y: osc.y,
       ...
    }
```

That is because it is the singleton that owns the statistics, not the instance attached to it.

There are many occurances of instance reaching into singleton and modifying it, which is code in the wrong "object".